mlflow/system_metrics/system_metrics_monitor.py

```diff
@@ -0,0 +1,133 @@
+        )
+
+        self._run_id = run_id
+        self.mlflow_logger = BatchMetricsLogger(self._run_id)
+        self._shutdown_event = threading.Event()
+        self._process = None
+
+    def start(self):
+        """Start monitoring system metrics."""
+        try:
+            self._process = threading.Thread(
+                target=self.monitor,
+                daemon=True,
+                name="SystemMetricsMonitor",
+            )
+            self._process.start()
+            _logger.info("MLflow: started monitoring system metrics.")
+        except Exception as e:
+            _logger.warning(f"MLflow: failed to start monitoring system metrics: {e}")
+            self._process = None
+
+    def monitor(self):
+        """Main monitoring loop, which consistently collect and log system metrics."""
+        from mlflow.tracking.fluent import get_run
+
+        while not self._shutdown_event.is_set():
+            for _ in range(self.samples_before_logging):
+                self.collect_metrics()
+                self._shutdown_event.wait(self.sampling_interval)
+                try:
+                    # Get the MLflow run to check if the run is not RUNNING.
+                    run = get_run(self._run_id)
+                except Exception as e:
+                    _logger.warning(f"MLflow: failed to get mlflow run: {e}.")
+                    return
+                if run.info.status != "RUNNING" or self._shutdown_event.is_set():
+                    # If the mlflow run is terminated or receives the shutdown signal, stop
+                    # monitoring.
+                    return
+            metrics = self.aggregate_metrics()
+            try:
+                self.publish_metrics(metrics)
+            except Exception as e:
+                _logger.warning(
+                    f"MLflow: failed to log system metrics: {e}, this is expected if the "
+                    "experiment/run is already terminated."
+                )
+                return
+
+    def collect_metrics(self):
+        """Collect system metrics."""
+        metrics = {}
+        for monitor in self.monitors:
+            monitor.collect_metrics()
+            metrics.update(monitor._metrics)
+        return metrics
+
+    def aggregate_metrics(self):
+        """Aggregate collected metrics."""
+        metrics = {}
+        for monitor in self.monitors:
+            metrics.update(monitor.aggregate_metrics())
+        return metrics
+
+    def publish_metrics(self, metrics):
+        """Log collected metrics to MLflow."""
+        self.mlflow_logger.record_metrics(metrics)
+        for monitor in self.monitors:
+            monitor.clear_metrics()
+
+    def finish(self):
+        """Stop monitoring system metrics."""
+        if self._process is None:
+            return
+        _logger.info("MLflow: stopping system metrics monitoring...")
+        self._shutdown_event.set()
+        try:
+            self._process.join()
```

Comment:
```suggestion
        self._shutdown_event.set()
        try:
            self._process.join()
            _logger.info("MLflow: stopped system metrics monitoring...")
```

Can we use `stopped` instead of `stopping`? `stopping` might make users wonder "did it stop or not?".

Reply:
reworded!